### DISCUSSION TITLE
Removing delay of presenting BottomSheetModal - Where to call `present` method?

### DISCUSSION THREAD

**Discussion by theMasix** [gorhom/react-native-bottom-sheet#426] [Q&A] [answered]
Hello.
Since BottomSheetModal needs to be presented before expanding, 2 approaches might be applied to show bottom sheet:
1. configuring `snapPoints` and `index` to be expanded by default, then after calling `present` method, bottom sheet will pop up. The issue is if we omit `expand` method and only call `present` to show bottom sheet, then it has some delay for mounting process.
2. configuring bottom sheet not to be shown on mount, then calling `present` on screen where bottom sheet component exists, after mounting, we could show the bottom sheet with `expand` method instantly.

My question is, since we don't want have any delay before showing bottom sheet and we have to use BottomSheetModal (we have some tab navigators that should be under bottom sheet), then where should we call `present`?
Should we call `present` method in `useFocusEffect` (we use `react-navigation` to handle screens)? Should it be called in `useLayoutEffect` or what?

In first approach where we call `present` in `useFocusEffect` then the bottom sheet `ref` might not be passed to bottom sheet component and `present` method might be null. We could handle it with `setTimeout` in `useFocusEffect` but it's ugly and isn't the correct approach I think.

Any Idea? Thanks

**Comment by pouyarezvani** [comment]
@theMasix I see your post is from May 2nd. Have you found any solution to your delay issue? We are experiencing the same issue. There is a bit of delay before BottomSheetModal is presented. 
@gorhom Would you be able to see into this if possible, please? 
Thanks.

**Reply by theMasix** [reply]
Unfortunately No. Although the delay is much shorter in production builds, but we haven't reached a good solution to solve it completely yet. The library is cool but its community is not active :(

**Reply by gorhom** [reply]
Could you test on  ‘v4’ ?

**Reply by jparkrr** [reply]
I'm also trying to make the sheet visible on mount.

@gorhom I've tested in on v4, and am running into https://github.com/gorhom/react-native-bottom-sheet/issues/505

I presume it's calculating some values for the layout before it gets shown? Do you know much we can expect this to be reduced when v4 gets released?

Thanks for your work!

**Reply by theMasix** [reply]
I noticed that the #505 is fixed now. Is the delay fixed in `v4` or not yet? @jparkrr

**Comment by theMasix** [comment, marked as answer]
I completely solved the issue by using [`gorhom/react-native-portal`](https://github.com/gorhom/react-native-portal) instead of `BottomSheetModal` component.

don't know why the `BottomSheetModal` has some delay on mount!

**Reply by theMasix** [reply]
@jparkrr @gorhom @pouyarezvani

**Reply by pouyarezvani** [reply]
Can you provide an expo snack that fixes this problem?

**Reply by theMasix** [reply]
It's so simple that I think it doesn't need snack sample.

1. Wrap your `Navigation` component with `PortalProvider` instead of `BottomSheetModalProvider`.
```jsx
  <PortalProvider>
         <Navigation />
  </PortalProvider>
```

2. Replace `BottomSheetModal` with `BottomSheet` component and wrap it with `Portal` component.
```jsx
      <BottomSheetModal
        ref={ref}
        index={0}
        snapPoints={snapPoints}
        bottomInset={0}
        ...
```

to

```jsx
    <Portal>
      <BottomSheet
        ref={ref}
        index={-1}
        snapPoints={snapPoints}
        bottomInset={0}
        ...
```

You may need to update your `index` property and also `BottomSheetBackdrop`s `appearsOnIndex` and `disappearsOnIndex` property if you use it as well.

**Reply by pouyarezvani** [reply]
How much delay did you have before? How long was it? 
In my experience currently there is a .5 second delay before the modal shows up.

**Reply by theMasix** [reply]
> How much delay did you have before? How long was it?

depends of size of the screen, between 0.5 to 1sec

**Reply by pouyarezvani** [reply]
@theMasix The problem with using BottomSheet instead of `BottomSheetModal` is that you cant dismiss the `BottomSheet`. Are you able to dismiss it still after you switched to `BottomSheet`?
Also, are you using `ref.current?.expand( 1 )`?
I have also lost the animation on when it opens.

That's why I was hoping to see a snack of this, to test it out.

**Reply by theMasix** [reply]
Yeah I didn't have any issue to dismiss. But I have other issues such as not closing the bottom sheet completely right now.

I don't use `expand(1)`. What's your ```snapPoints``` value?

I hope I can create an snack for you tomorrow.

**Reply by pouyarezvani** [reply]
I changed the `expand(1)` to `ref.current?.snapTo( 2, 500 )`
And My `snapPoints` are `snapPoints={[ 0,'25%', '50%' ]}`
And my Index is `index={-1}`.

^ This is not working as expected behavior

**Reply by pouyarezvani** [reply]
@tgecty any luck on the snack?

**Reply by bryanprimus** [reply]
still an issue. the delay is annoying. i believe this happen because the BottomSheetModal isn't mounted by default. maybe it takes time to be mounted the first time?

**Reply by Karthik-B-06** [reply]
Hey, for me personally, the Bottomsheet Modal is presented properly, but when closed, I am unable to quickly interact with the screen. There is a delay. Is anyone facing this issue?

**Reply by myungjael** [reply]
@Karthik-B-06 After closing the bottomSheetModal, you need to open another bottomSheetModal quickly, but there is a delay. In the end, there was no solution, so I replaced it with a normal Modal.
